server/implementation/src/main/java/io/smallrye/graphql/bootstrap/Bootstrap.java

```diff
@@ -184,7 +189,32 @@ private void generateGraphQLSchema() {
         Map<String, Jsonb> overrides = eventEmitter.fireOverrideJsonbConfig();
         JsonInputRegistry.override(overrides);
 
-        this.graphQLSchema = schemaBuilder.build();
+        if (Config.get().isFederationEnabled()) {
+            log.enableFederation();
+            GraphQLSchema rawSchema = schemaBuilder.build();
+            this.graphQLSchema = Federation.transform(rawSchema)
+                    .fetchEntities(new FederationDataFetcher(rawSchema.getQueryType(), rawSchema.getCodeRegistry()))
+                    .resolveEntityType(fetchEntityType())
+                    .build();
+        } else {
+            this.graphQLSchema = schemaBuilder.build();
+        }
+    }
+
+    private TypeResolver fetchEntityType() {
+        return env -> {
+            Object src = env.getObject();
+            if (src == null) {
+                return null;
+            }
+            Name annotation = src.getClass().getAnnotation(Name.class);
+            String typeName = (annotation == null) ? src.getClass().getSimpleName() : annotation.value();
+            GraphQLObjectType result = env.getSchema().getObjectType(typeName);
+            if (result == null) {
+                throw new RuntimeException("can't resolve federated entity type " + src.getClass().getName());
+            }
```

Comment:
This needs to happen with Jandex, when building the model, so the model might have to be enhanced. Bootstrap (and any other class used in runtime) will then use the model.

Reply:
If we do the same in InterfaceResolver and UnionResolver, we need to fix those too. I'll check out your PR branch and see if this is going to be possible. We might not be able to since we are using a library. I'll get back to you a.s.a.p

Reply:
Do you consider this PR as incomplete or rather just imperfect? If the latter, we could, maybe, merge this as it is and fix these shortcomings in a separate PR.

Reply:
I think it's complete, we can merge and then follow up, but just give me some time, I'll merge either way later this week. Is that ok ?

Reply:
sure!